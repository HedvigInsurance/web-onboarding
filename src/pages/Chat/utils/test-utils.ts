import { OfferData } from 'containers/OfferContainer'
import { MockedResponse } from 'react-apollo/test-links'
import { InsuranceType } from 'utils/insuranceDomainUtils'
import {
  CREATE_OFFER_MUTATION,
  CreateOfferMutationVariables,
} from '../containers/CreateOfferContainer'
import { ChatStep, Insurer, State as ChatState } from '../state'

export const mockState = (): ChatState => ({
  nameAge: {
    firstName: 'John',
    lastName: 'Doe',
    age: 22,
  },
  livingSituation: {
    postalNumber: '123 45',
    streetAddress: 'Storgatan 1',
    insuranceType: InsuranceType.RENT,
    size: 37,
    numberOfPeople: 1,
  },
  isStudent: false,
  currentInsurance: {
    currentInsurer: Insurer.FOLKSAM,
    hasCurrentInsurance: true,
  },
  currentStep: ChatStep.SHOW_OFFER,
  initialVisibleSteps: [],
  visibleSteps: [],
})

export const mockOffer: OfferData = {
  insurance: {
    address: 'TestvÃ¤gen 1',
    monthlyCost: 99,
    insuredAtOtherCompany: false,
    type: InsuranceType.RENT,
    postalNumber: '12345',
    personsInHousehold: 1,
    currentInsurerName: Insurer.FOLKSAM,
  },
  member: {
    firstName: 'Test',
    lastName: 'Testerson',
  },
}

export const createCreateOfferMutationMock = (
  isStudent: boolean = false,
): MockedResponse[] => [
  {
    request: {
      query: CREATE_OFFER_MUTATION,
      // tslint:disable-next-line no-object-literal-type-assertion
      variables: {
        firstName: mockState().nameAge.firstName,
        lastName: mockState().nameAge.lastName,
        age: mockState().nameAge.age,
        address: mockState().livingSituation.streetAddress,
        postalNumber: '12345',
        squareMeters: mockState().livingSituation.size,
        insuranceType: isStudent
          ? InsuranceType.STUDENT_RENT
          : InsuranceType.RENT,
        personsInHousehold: mockState().livingSituation.numberOfPeople,
        previousInsurer: mockState().currentInsurance.currentInsurer,
      } as CreateOfferMutationVariables,
    },
    result: {
      data: { createOffer: 'abc123' },
    },
  },
]
