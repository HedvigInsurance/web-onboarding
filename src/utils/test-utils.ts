export const mockNetworkWait = (delay: number = 0) =>
  new Promise<void>((resolve) => setTimeout(() => resolve(), delay))
