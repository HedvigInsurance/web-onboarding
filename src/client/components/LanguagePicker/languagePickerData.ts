import { Market, LocalePath } from 'src/client/components/utils/CurrentLocale'

const LANGUAGE_TITLES = {
  english: 'English',
  swedish: 'Svenska',
  norwegian: 'Norsk',
  danish: 'Dansk',
}

const SHORT_LANGUAGE_TITLES = {
  english: 'En',
  swedish: 'Sv',
  norwegian: 'No',
  danish: 'Da',
}

type Options = {
  linkTo: LocalePath
  title: string
  shortTitle: string
}[]

type LanguagePickerData = Record<Market, Options>

export const languagePickerData: LanguagePickerData = {
  SE: [
    {
      linkTo: 'se',
      title: LANGUAGE_TITLES.swedish,
      shortTitle: SHORT_LANGUAGE_TITLES.swedish,
    },
    {
      linkTo: 'se-en',
      title: LANGUAGE_TITLES.english,
      shortTitle: SHORT_LANGUAGE_TITLES.english,
    },
  ],
  NO: [
    {
      linkTo: 'no',
      title: LANGUAGE_TITLES.norwegian,
      shortTitle: SHORT_LANGUAGE_TITLES.norwegian,
    },
    {
      linkTo: 'no-en',
      title: LANGUAGE_TITLES.english,
      shortTitle: SHORT_LANGUAGE_TITLES.english,
    },
  ],
  DK: [
    {
      linkTo: 'dk',
      title: LANGUAGE_TITLES.danish,
      shortTitle: SHORT_LANGUAGE_TITLES.danish,
    },
    {
      linkTo: 'dk-en',
      title: LANGUAGE_TITLES.english,
      shortTitle: SHORT_LANGUAGE_TITLES.english,
    },
  ],
}
