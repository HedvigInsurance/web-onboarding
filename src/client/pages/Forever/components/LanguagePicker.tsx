import styled from '@emotion/styled'
import { colorsV3 } from '@hedviginsurance/brand'
import { Market, useMarket } from 'components/utils/CurrentLocale'
import React from 'react'
import { useParams } from 'react-router'
import { NavLink } from 'react-router-dom'

const LanguagePickerWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
`
const Divider = styled.div`
  width: 1px;
  margin-left: 2px;
  margin-right: 2px;
  border-left: 1px solid #777;
`
const Language = styled(NavLink)`
  text-decoration: none;
  color: #777;

  &.active {
    color: ${colorsV3.white};
  }

  :hover,
  :focus {
    color: ${colorsV3.white};
  }
`

export const LanguagePicker: React.FC = () => {
  const market = useMarket()
  const urlParams: { locale: string; code?: string } = useParams()
  const code = urlParams.code ? urlParams.code : ''

  if (market === Market.No) {
    return (
      <LanguagePickerWrapper>
        <Language to={`/no/forever/${code}`}>No</Language>
        <Divider />
        <Language to={`/no-en/forever/${code}`}>En</Language>
      </LanguagePickerWrapper>
    )
  }
  return (
    <LanguagePickerWrapper>
      <Language to={`/se/forever/${code}`}>Sv</Language>
      <Divider />
      <Language to={`/se-en/forever/${code}`}>En</Language>
    </LanguagePickerWrapper>
  )
}
