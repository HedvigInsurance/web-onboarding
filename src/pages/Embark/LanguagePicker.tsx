import styled from '@emotion/styled'
import { colorsV2 } from '@hedviginsurance/brand'
import { LanguageIcon } from 'components/icons/LanguageIcon'
import { Market, useMarket } from 'components/utils/CurrentLocale'
import { motion } from 'framer-motion'
import * as React from 'react'
import { Link } from 'react-router-dom'

const Wrapper = styled.div`
  position: relative;
`

const LanguageDropdownButton = styled.button`
  background: none;
  padding: 0;
  margin: 0;
  border: none;
  cursor: pointer;
  outline: 0;
`

const StyledLink = styled(Link)`
  outline: 0;
  color: ${colorsV2.black};
  text-decoration: none;
  padding: 10px 20px;
  transition: all 250ms;
  background-color: white;
  font-size: 0.9rem;

  :active {
    background-color: ${colorsV2.lightgray};
  }
`

const Divider = styled.span`
  height: 1px;
  width: 100%;
  background-color: ${colorsV2.lightgray};
`

const Dropdown = styled.div`
  background-color: white;
  border-radius: 5px;
  position: fixed;
  transform: translateX(-40%) translateY(10px);
  display: flex;
  flex-direction: column;
  overflow: hidden;
  z-index: 999;
`

export const LanguagePicker = () => {
  const [isOpen, setIsOpen] = React.useState(false)
  const market = useMarket()

  return (
    <Wrapper>
      <LanguageDropdownButton onClick={() => setIsOpen(!isOpen)}>
        <LanguageIcon />
      </LanguageDropdownButton>
      <motion.div
        initial={{ opacity: 0, y: -20 }}
        animate={
          isOpen
            ? { opacity: 1, y: 0, pointerEvents: 'auto', userSelect: 'auto' }
            : { opacity: 0, y: -20, pointerEvents: 'none', userSelect: 'none' }
        }
        transition={{ type: 'spring', stiffness: 400, damping: 100 }}
      >
        <Dropdown>
          {market === Market.Se && (
            <>
              <StyledLink to="/new-member">Svenska</StyledLink>
              <Divider />
              <StyledLink to="/en/new-member">English</StyledLink>
            </>
          )}
          {market === Market.No && (
            <>
              <StyledLink to="/no/new-member">Norsk</StyledLink>
              <Divider />
              <StyledLink to="/no-en/new-member">English</StyledLink>
            </>
          )}
        </Dropdown>
      </motion.div>
    </Wrapper>
  )
}
