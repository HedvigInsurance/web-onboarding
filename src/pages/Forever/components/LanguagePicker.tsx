import styled from '@emotion/styled'
import { colorsV3 } from '@hedviginsurance/brand'
import { Market, useMarket } from 'components/utils/CurrentLocale'
import React from 'react'
import { RouteComponentProps, withRouter } from 'react-router'
import { NavLink } from 'react-router-dom'

const LanguagePickerWrapper = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: center;
`
const Divider = styled.div`
  width: 1px;
  margin-left: 2px;
  margin-right: 2px;
  border-left: 1px solid #777;
`
const Language = styled(NavLink)`
  text-decoration: none;
  color: #777;

  &.active {
    color: ${colorsV3.white};
  }

  :hover,
  :focus {
    color: ${colorsV3.white};
  }
`

export const LanguagePickerComponent: React.FC<{ code?: string }> = ({
  code,
}) => {
  const market = useMarket()

  if (market === Market.No) {
    return (
      <LanguagePickerWrapper>
        <Language to={`/no/forever/${code ?? ''}`}>No</Language>
        <Divider />
        <Language to={`/no-en/forever/${code ?? ''}`}>En</Language>
      </LanguagePickerWrapper>
    )
  }
  return (
    <LanguagePickerWrapper>
      <Language to={`/se/forever/${code ?? ''}`}>Sv</Language>
      <Divider />
      <Language to={`/se-en/forever/${code ?? ''}`}>En</Language>
    </LanguagePickerWrapper>
  )
}

const withCodeFromRouteMatch = (
  Component: React.ComponentType<{ code?: string }>,
) => (props: RouteComponentProps<{ code?: string }>) => {
  return <Component code={props.match.params.code} />
}

export const LanguagePicker = withRouter(
  withCodeFromRouteMatch(LanguagePickerComponent),
)
