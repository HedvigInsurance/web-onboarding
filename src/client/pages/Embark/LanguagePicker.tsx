import styled from '@emotion/styled'
import { colorsV2 } from '@hedviginsurance/brand'
import { motion } from 'framer-motion'
import React from 'react'
import { Link } from 'react-router-dom'
import { useMarket } from 'components/utils/CurrentLocale'
import { LanguageIcon } from 'components/icons/LanguageIcon'
import { languagePickerData } from './languagePickerData'

const Wrapper = styled.div`
  position: relative;
`
const Option = styled.div`
  display: flex;
  flex-direction: column;
`

const LanguageDropdownButton = styled.button`
  background: none;
  padding: 0;
  margin: 0;
  border: none;
  cursor: pointer;
  outline: 0;
`

const StyledLink = styled(Link)`
  outline: 0;
  color: ${colorsV2.black};
  text-decoration: none;
  padding: 10px 20px;
  transition: all 250ms;
  font-size: 0.9rem;

  :active {
    background-color: ${colorsV2.lightgray};
  }
`

const Divider = styled.span`
  height: 1px;
  width: 100%;
  background-color: ${colorsV2.lightgray};
`

const Dropdown = styled.div`
  height: 80px;
  background-color: white;
  border-radius: 5px;
  position: absolute;
  right: 0;
  transform: translateY(10px);
  display: flex;
  flex-direction: column;
  justify-content: space-around;
  overflow: hidden;
  z-index: 9999;
  box-shadow: 0 3px 5px rgba(0, 0, 0, 0.1);
`

export const LanguagePicker: React.FC<{ path?: string }> = ({
  path = '/new-member',
}) => {
  const [isOpen, setIsOpen] = React.useState(false)
  const market = useMarket()

  React.useEffect(() => {
    const clickListener = () => {
      setIsOpen(false)
    }
    document.body.addEventListener('click', clickListener)

    return () => document.body.removeEventListener('click', clickListener)
  }, [])

  return (
    <Wrapper>
      <LanguageDropdownButton onClick={() => setIsOpen(!isOpen)}>
        <LanguageIcon />
      </LanguageDropdownButton>
      <motion.div
        initial={{ opacity: 0, y: -20 }}
        variants={{
          open: { opacity: 1, y: 0 },
          closed: {
            opacity: 0,
            y: -20,
          },
        }}
        animate={isOpen ? 'open' : 'closed'}
        style={{
          pointerEvents: isOpen ? 'auto' : 'none',
          userSelect: isOpen ? 'auto' : 'none',
        }}
        transition={{ type: 'spring', stiffness: 400, damping: 100 }}
      >
        <Dropdown>
          {languagePickerData[market].map(({ linkTo, title }, index) => (
            <Option key={linkTo}>
              <StyledLink to={`/${linkTo}${path}`}>{title}</StyledLink>
              {index < languagePickerData[market].length - 1 && <Divider />}
            </Option>
          ))}
        </Dropdown>
      </motion.div>
    </Wrapper>
  )
}
