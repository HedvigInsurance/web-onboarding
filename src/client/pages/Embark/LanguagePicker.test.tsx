import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import React from 'react'
import { act } from 'react-dom/test-utils'
import { MemoryRouter } from 'react-router-dom'
import { LanguagePicker } from './LanguagePicker'

jest.mock('../../apolloClient', () => ({}))

it('renders SE without ðŸ’¥', () => {
  const wrapper = mount(
    <MemoryRouter initialEntries={['/se/new-member']}>
      <LanguagePicker />
    </MemoryRouter>,
  )

  expect(wrapper.find('Link[to*="/se/new-member"]')).toHaveLength(1)
  expect(wrapper.find('Link[to*="/se-en/new-member"]')).toHaveLength(1)
  expect(wrapper.find('Link[to*="/no/new-member"]')).toHaveLength(0)
  expect(wrapper.find('Link[to*="/no-en/new-member"]')).toHaveLength(0)

  expect(toJson(wrapper.find(LanguagePicker))).toMatchSnapshot()
})

it('renders NO without ðŸ’¥', () => {
  const wrapper = mount(
    <MemoryRouter initialEntries={['/no/new-member']}>
      <LanguagePicker />
    </MemoryRouter>,
  )

  expect(wrapper.find('Link[to*="/no/new-member"]')).toHaveLength(1)
  expect(wrapper.find('Link[to*="/no-en/new-member"]')).toHaveLength(1)
  expect(wrapper.find('Link[to*="/se/new-member"]')).toHaveLength(0)
  expect(wrapper.find('Link[to*="/se-en/new-member"]')).toHaveLength(0)

  expect(toJson(wrapper.find(LanguagePicker))).toMatchSnapshot()
})

it('renders DK without ðŸ’¥', () => {
  const wrapper = mount(
    <MemoryRouter initialEntries={['/dk/new-member']}>
      <LanguagePicker />
    </MemoryRouter>,
  )

  expect(wrapper.find('Link')).toHaveLength(2)
  expect(wrapper.find('Link[to*="/dk/new-member"]')).toHaveLength(1)
  expect(wrapper.find('Link[to*="/dk-en/new-member"]')).toHaveLength(1)

  expect(toJson(wrapper.find(LanguagePicker))).toMatchSnapshot()
})

it('opens when clicking icon', () => {
  const wrapper = mount(
    <MemoryRouter initialEntries={['/se/new-member']}>
      <LanguagePicker />
    </MemoryRouter>,
  )

  expect(wrapper.find('ForwardRef(MotionComponent)').prop('animate')).toBe(
    'closed',
  )
  act(() => {
    wrapper.find('LanguageDropdownButton').simulate('click')
  })
  wrapper.update()

  expect(wrapper.find('ForwardRef(MotionComponent)').prop('animate')).toBe(
    'open',
  )
})
