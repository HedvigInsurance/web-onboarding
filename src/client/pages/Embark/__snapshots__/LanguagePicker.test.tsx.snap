// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Danish language picker renders without ðŸ’¥ 1`] = `
<div
  class="css-1prqaxz-Wrapper e1g7xtvp0"
>
  <div
    class="css-14fi736-ActiveOption e1g7xtvp1"
    color="black"
  >
    Da
  </div>
  <span
    class="css-1tysqfo-Divider e1g7xtvp3"
    color="black"
  />
  <a
    class="css-s5uyen-LinkOption e1g7xtvp2"
    color="black"
    href="/dk-en/new-member"
  >
    En
  </a>
</div>
`;

exports[`Norwegian language picker renders without ðŸ’¥ 1`] = `
<div
  class="css-1prqaxz-Wrapper e1g7xtvp0"
>
  <div
    class="css-14fi736-ActiveOption e1g7xtvp1"
    color="black"
  >
    No
  </div>
  <span
    class="css-1tysqfo-Divider e1g7xtvp3"
    color="black"
  />
  <a
    class="css-s5uyen-LinkOption e1g7xtvp2"
    color="black"
    href="/no-en/new-member"
  >
    En
  </a>
</div>
`;

exports[`Swedish language picker renders without ðŸ’¥ 1`] = `
<div
  class="css-1prqaxz-Wrapper e1g7xtvp0"
>
  <div
    class="css-14fi736-ActiveOption e1g7xtvp1"
    color="black"
  >
    Sv
  </div>
  <span
    class="css-1tysqfo-Divider e1g7xtvp3"
    color="black"
  />
  <a
    class="css-s5uyen-LinkOption e1g7xtvp2"
    color="black"
    href="/se-en/new-member"
  >
    En
  </a>
</div>
`;
